### ISSUE TITLE
BUILDKITE_AGENT_SPAWN > 1 sets no unique agent tags

### ISSUE THREAD

**Issue by willhughes-au** [buildkite/agent#1741]
**Is your feature request related to a problem? Please describe.**

We use Buildkite from the official Docker container.  We use the BUILDKITE_AGENT_SPAWN environment variable to start up multiple agent processes in the one container. 

There is a need for us to be able to target jobs to a very specific agent. 

When using the API to query for the agents, we can see the agent's name and id. These are unique values for us, and would be appropriate to set as the target for a job. 
However, the agent does not have either of these as a tag, and so jobs cannot be targeted to a specific agent. 

**Describe the solution you'd like**

We would like to see some unique tags added to each Buildkite Agent process. 
For example a tag with it's own name, or id. 

**Describe alternatives you've considered**

Buildkite Support has suggested modifying the entrypoint scripts for the container, however as Buildkite-Agent uses the `tini` init script - we would lose any signal handling that tini provides.  

Moving to something like s6 or another init replacement would be a major departure from the current container and make it more difficult to track.

Moving to a single buildkite-agent per container would increase costs/complexity of managing our Buildkite Agent infrastructure. 

**Additional context**
n/a

**Comment by moskyb** [comment]
kia ora @willhughes-au! can you elaborate a bit on the use case where you need to specifically target one particular spawned agent? theoretically, all of the agents instances spawned by one agent should have the same config, and should be equivalent, but that obviously isn't the case here.

if you'd rather not talk about it in the open, then feel free to DM me on the buildkite slack, or send me an email at ben.m@buildkite.com.

**Comment by willhughes-au** [comment]
Hi @moskyb 

> "theoretically, all of the agents instances spawned by one agent should have the same config"  

Yes, we agree -- and in most scenarios that's true.   

We have run across this a few times, however, and in each case managed to find some work-around. 

This time it was because we have agents running steps in a bunch of pipelines which generate *a lot* of files (`node_modules` directory), and this is exhausting the number of inodes available on the container disk.

What we were looking for is some way of cleaning up/removing the pipeline working directory on agents, when the pipeline is completed. That doesn't appear to exist.  
My thought was to just run a periodic pipeline, which generates a job for every spawned agent, and that job would then remove all the pipeline working directories for that spawned agent. 

Without the ability to target a specific spawned agent, we can't guarantee that each agent will actually run the task. i.e we might have all four cleanup tasks for `hostname=X` run on the same agent. 
We also can't just cleanup the working directory for other other spawned agents on the same host, as this will run into race conditions where it's removing in-use folders.

In short, being able to target a specific agent would allow us to perform certain actions that we can't otherwise without it.

**Comment by moskyb** [comment]
that's a lot of really helpul context, thank you!

> which generate a lot of files (node_modules directory)

ah, that old chestnut 😅

we don't have any plans to ship features that makes targeting specific agents easier - it goes against our cattle-vs-pets view of buildkite agent fleets, but i understand that this is making things a little tricky for you.

in this case, i think a similar goal be achievable by adding a pre-exit hook that runs `git clean -ffxdq` after a job run. This is what happens the next time a job runs on this agent and pipeline anyway; having it execute in a hook just brings it forward in the process so node_modules isn't sticking around on your machine taking up inodes between runs.

**Comment by moskyb** [comment]
oh, `-ffxdq` on git clean means:
```
// -ff: delete files and directories, including untracked nested git repositories
// -x: don't use .gitignore rules
// -d: recurse into untracked directories (eg node_modules)
// -q: quiet, only report errors
```

**Comment by willhughes-au** [comment]
> it goes against our cattle-vs-pets view of buildkite agent fleets

I think we're in pretty close agreement in the general case. 

However to push the analogy to perhaps breaking point: even on farms, you sometimes need to deal with members of the herd/flock/$collective_noun individually. It doesn't necessarily make them pets. 

The pre-execute hook might work, but it might cause other conflicts. 
We've got some other work-arounds for this particular scenario which lets us keep going (restart the agents many times a day). 

All in all they all seem like less than ideal solutions.

**Comment by pda** [comment]
Howdy @willhughes-au 👋🏼

Let me see if I understand the issue:

- You're running `buildkite-agent` within docker container(s), using `--spawn` to run multiple agents in the same process/container.
- These agents handle build jobs for many different pipelines.
- A new checkout directory is created [for each spawn+pipeline combo](https://github.com/buildkite/agent/blob/31c30294d8908992b667c15f528b102287d6211e/bootstrap/bootstrap.go#L519-L520).
- After running jobs from many different pipelines, many different checkout directories are left behind with the full working directory as left by the most recent build job. e.g. `--spawn=4` handling 10 different pipelines could accumulate 40 working directories
- Eventually, disk space and/or inodes are exhausted.

The fact that each build job leaves behind its working directory contents (and that each spawn/pipeline combo gets a new directory) only for the next job in the spawn/pipeline combo to `git clean -ffxdq` it doesn't feel like a useful feature, _unless_ you sometimes need to manually dive in and inspect the result of a build after it's finished (but before another job wipes it).

In that case, I would definitely recommend `git clean $BUILDKITE_GIT_CLEAN_FLAGS` in an agent `pre-exit` hook (e.g. `/buildkite/hooks/pre-exit`, depending on your `hooks-path` config in the container). That will leave behind the `.git` directory so the next job remains fast, but will eliminate the working tree which is likely what's using all your inodes.

I actually wonder if this should become the default, or at least have a simple flag to opt in 🤔

Regarding tagging the agent by its spawn index, I think we'd only be interested in pursuing that if it solved a problem that couldn't be solves via agent hooks etc. So I'd be keen to hear if the `pre-exit` solution is viable — if so I think it will end up much simpler and more reliable than pointing garbage-collection builds at all known agents.

**Comment by pda** [comment]
👋🏼 Hey @willhughes-au, have you had a chance to see if the `pre-exit` approach would work for you?

**Comment by willhughes-au** [comment]
Hi, yes we've added the pre-exit hooks, and also done some other things to fix this specific case. 

But as mentioned, we've hit this repeatedly, and had to work around it in different ways. 

If we were not using spawn at all, then this wouldn't be an issue.  
We use spawn because it simplifies other things around scheduling of agents.  

I guess we are stuck having to trade off simplified management of agents (being able to target one agent) vs simplified scheduling of agents (using spawn). 

>we'd only be interested in pursuing that if it solved a problem that couldn't be solves via agent hooks etc

In that case, I guess I'll close the issue.
